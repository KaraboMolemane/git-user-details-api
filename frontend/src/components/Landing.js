import { useState, useEffect } from 'react';
import 'react-data-grid/lib/styles.css';
//import DataGrid from 'react-data-grid';
import * as React from 'react';
import Box from '@mui/material/Box';
import { DataGrid } from '@mui/x-data-grid';
// import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
// import { faBook } from '@fortawesome/free-solid-svg-icons';
import { Link  } from 'react-router-dom';


function Landing() {
  //declare state(s)
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [items, setItems] = useState([]);
  const [searchText, setSearchText] = useState('');

  // const columns = [
  //   { key: 'id', name: 'ID' },
  //   { key: 'title', name: 'Title' }
  // ];
  
  // const rows = [
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },
  //   { id: 1, title: 'Demo' },
  //   { id: 0, title: 'Example' },


  // ];

  const columns = [
    { field: 'id', 
      headerName: 'ID', 
      width: 150,
      flex: 1
    },
    {
      field: 'login',
      headerName: 'User name',
      width: 350,
      flex: 2

    },
    {
        field: 'url',
        headerName: 'URL',
        width: 350,
        flex: 2
  
    },
    {
      field: 'details',
      headerName: 'User details',
      width: 150,
      flex: 1,
      cellClassName: 'user-details--cell',
    },
  ];
  
  // const rows = [
  //   { id: 1, login: 'Snow', details: 'Jon',},
  //   { id: 2, login: 'Lannister', details: 'Cersei'},
  //   { id: 3, login: 'Lannister', details: 'Jaime'},
  //   { id: 4, login: 'Stark', details: 'Arya'},
  //   { id: 5, login: 'Targaryen', details: 'Daenerys'},
  //   { id: 6, login: 'Melisandre', details: null },
  //   { id: 7, login: 'Clifford', details: 'Ferrara' },
  //   { id: 8, login: 'Frances', details: 'Rossini'},
  //   { id: 9, login: 'Roxie', details: 'Harvey'},
  // ];

  let rows = [];
  items.forEach(element => {
      rows.push({
      id: element.id, 
      login: element.login, 
      url: element.url,
      details: 'See more'
    })
  })


//Implement event for cellClick
// const handleRowClick = (params) => {
//   console.log('Well, you clicked Row:', params);
// };

const handleCellClick = (params) => {
  console.log('Well, you clicked the Cell:', params);
  // Only reroute if the user clicks on the the details column cell 
  // if(params.field === 'details') console.log('Route:', params.row.login);
   if(params.field === 'details') window.location.href = '/user-details?userurl='+params.row.url;
};

  useEffect(() => {
    if(searchText !== ''){

      //Do the API call
      fetch('/search-users/' + searchText)
      //fetch('https://api.github.com/search/users?q=karabo in:name type:user')
         .then(res => res.json())
        //.then(res => console.log('res:', res))
        .then(
          (result) => {
            setIsLoaded(true);
            setItems(result.items);
          },
          (error) => {
            setIsLoaded(true);
            setError(error);
          }
        )
    }

  }, [searchText])

  let results = '';
  if (error) {
    results =  <div>Error: {error.message}</div>;
  } else if (!isLoaded) {
    results = <div>{searchText === ''? 'Enter text above to search.' : 'Loading...'}</div>;
  } else {

    console.log('items:', items);

  }


  return (
    <div className="container my-5">
    <div className="p-5 text-center bg-body-tertiary rounded-3">
        <br />
        <h1 className="text-body-emphasis">Search GitHub User App</h1>
        <p className="col-lg-8 mx-auto fs-5 text-muted">
          This app allows you to search for GitHub users and see the following:
        </p>
        <ul className='list-unstyled'>
          <li>User details - including some of their repos, their profile picture, bio, etc.</li>
          <li>Repo details - including last commit date, creation date, description, etc.</li>
        </ul>
        <div className="bd-example">
      <form>
        <div className="w-50 p-3" style={{marginLeft: '30%'}} >
          <input type="text" className="form-control d-inline-flex align-items-center" id="userInputWord" aria-describedby="emailHelp" placeholder='Type in text to search for users'/>          
        </div>       
      </form>
      <br />
      <div>{results}</div>      <br />
        <div className="d-inline-flex gap-2 mb-5">

            <button type="button" className="d-inline-flex align-items-center btn btn-primary btn-lg px-4 rounded-pill" data-bs-toggle="modal" data-bs-target="#exampleModalFullscreen" onClick={() => setSearchText(document.getElementById('userInputWord').value)}>
            Search
          </button>
        </div>
      
      </div>
      
    </div>
    
    {/* Modal for API results  */}
    <div className="modal fade" id="exampleModalFullscreen" tabindex="-1" aria-labelledby="exampleModalFullscreenLabel" aria-hidden="true">
  <div className="modal-dialog modal-fullscreen">
    <div className="modal-content">
      <div className="modal-header">
        <h1 className="modal-title fs-4" id="exampleModalFullscreenLabel">Search results for "{searchText}"</h1>
        <button type="button" className="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div className="modal-body">
        {/*<DataGrid columns={columns} rows={rows} />*/}
      <Box sx={{ 
        height: 400, 
        width: '100%', 
        '& .user-details--cell': {
          // make column look like a link
          color: 'blue',
          textDecoration: 'underline',
          cursor: 'pointer',
        }
        }}>
      <DataGrid
        //onRowClick={handleRowClick}
        onCellClick={handleCellClick}
        rows={rows}
        columns={columns}
        initialState={{
          pagination: {
            paginationModel: {
              pageSize: 5,
            },
          },
        }}
        pageSizeOptions={[5]}
        checkboxSelection
        disableRowSelectionOnClick
      />
    </Box>
      </div>
      <div className="modal-footer">
        <button type="button" className="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>


</div>
);
}

export default Landing;

/*
USEFUL Links 
https://mui.com/x/react-data-grid/style/
https://www.sitepoint.com/get-url-parameters-with-javascript/
https://stackoverflow.com/questions/45898789/react-router-pass-param-to-component
*/