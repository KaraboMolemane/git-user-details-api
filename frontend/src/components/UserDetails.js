import { useState, useEffect, useRef } from "react";
import "react-data-grid/lib/styles.css";
//import DataGrid from 'react-data-grid';
import * as React from "react";
import Box from "@mui/material/Box";
import { DataGrid } from "@mui/x-data-grid";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowUpRightFromSquare } from "@fortawesome/free-solid-svg-icons";
import Header from "./Header";
import { Modal, Button } from "react-bootstrap";

function UserDetails() {
  //declare state(s)
  const [error, setError] = useState(null);
  const [isLoaded, setIsLoaded] = useState(false);
  const [items, setItems] = useState([]);
  const [repos, setRepos] = useState([]);
  const [showModal, setShow] = useState(false);

  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);

  const queryString = window.location.search;
  console.log("queryString", queryString);
  const searchParams = new URLSearchParams(queryString);
  const userLogin = useRef(searchParams.get("login"));
  console.log("userLogin:", userLogin);

  const columns = [
    { field: "id", headerName: "ID", width: 150, flex: 1 },
    {
      field: "name",
      headerName: "Repo name",
      width: 350,
      flex: 2,
    },
    {
      field: "url",
      headerName: "URL",
      width: 350,
      flex: 2,
    },
    {
      field: "details",
      headerName: "Repo details",
      width: 150,
      flex: 1,
      cellClassName: "details--cell",
    },
  ];


  let rows = [];
  repos.forEach(element => {
      rows.push({
      id: element.id, 
      name: element.name, 
      url: element.html_url,
      details: 'See more'
    })
  })

  const handleCellClick = (params) => {
    console.log("Well, you clicked the Cell:", params);
    // Only reroute if the user clicks on the the details column cell
    // if(params.field === 'details') console.log('Route:', params.row.login);
    if (params.field === "details")
      window.location.href = "/user-details?login=" + params.row.login;
  };

  useEffect(() => {
    //Do the API call - user details
    fetch("/get-user-details/" + userLogin.current)
      .then((res) => res.json())
      //.then(res => console.log('res:', res))
      .then(
        (result) => {
          setIsLoaded(true);
          console.log("user details:", result);
          setItems(result);
        },
        (error) => {
          setIsLoaded(true);
          setError(error);
        }
      );

    //Do the API call - user repos
    fetch("/get-user-repos/" + userLogin.current)
      .then((res) => res.json())
      //.then(res => console.log('res:', res))
      .then(
        (result) => {
          setIsLoaded(true);
          console.log("user details:", result);
          setRepos(result);
        },
        (error) => {
          setIsLoaded(true);
          setError(error);
        }
      );
  }, []);

  return (
    <>
      <Header />
      <div className="col">
        <div className="card">
          <div className="row g-0">
            <div className="col-md-4">
              <img
                width="100%"
                height="250"
                src={items.avatar_url}
                alt="User_profile_pic"
              />
            </div>
            <div className="col-md-8">
              <div className="card-body">
                {/* User details */}
                <h5 className="card-title">Name: {items.name} </h5>
                <div className="bd-example-snippet bd-code-snippet">
                  <div className="bd-example">
                    <table className="table table-hover">
                      <tbody>
                        <tr className="table-light">
                          <th scope="row">Username:</th>
                          <td>{items.login}</td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Bio:</th>
                          <td>{items.bio}</td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">GitHub Url:</th>
                          <td>
                            <a href={items.html_url} target="_blank">
                              {items.html_url}
                            </a>{" "}
                            <FontAwesomeIcon icon={faArrowUpRightFromSquare} />
                          </td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Company:</th>
                          <td>{items.company} </td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Location:</th>
                          <td>{items.location}</td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Blog</th>
                          <td>
                            <a href={items.blog} target="_blank">
                              {items.blog}
                            </a>{" "}
                            <FontAwesomeIcon icon={faArrowUpRightFromSquare} />
                          </td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Email:</th>
                          <td>{items.email}</td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Followers:</th>
                          <td>{items.followers}</td>
                        </tr>
                        <tr className="table-light">
                          <th scope="row">Following:</th>
                          <td> {items.following}</td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
                <div>
                  {/* user repos */}
                  <h6>Repos</h6>
                  <Box
                    sx={{
                      height: 400,
                      width: "100%",
                      "& .details--cell": {
                        // make column look like a link
                        color: "blue",
                        textDecoration: "underline",
                        cursor: "pointer",
                      },
                    }}
                  >
                    <DataGrid
                      //onRowClick={handleRowClick}
                      onCellClick={handleCellClick}
                      rows={rows}
                      columns={columns}
                      initialState={{
                        pagination: {
                          paginationModel: {
                            pageSize: 5,
                          },
                        },
                      }}
                      pageSizeOptions={[5]}
                      checkboxSelection
                      disableRowSelectionOnClick
                    />
                  </Box>
                </div>
                <p className="card-text">
                  <small className="text-muted">
                    {items.name} joined GitHub on {items.created_at}
                  </small>
                </p>
                <a href="/" className="btn btn-primary">
                  Back to Homepage
                </a>
              </div>
            </div>
          </div>
        </div>
        {/* 
        React Modal for repo details  
        https://www.pluralsight.com/guides/how-to-trigger-modal-for-react-bootstrap
        */}
        <div
        className="d-flex align-items-center justify-content-center"
        style={{ height: "100vh" }}
      >
        <Button variant="primary" onClick={handleShow}>
          Launch demo modal
        </Button>
      </div>
      <Modal show={showModal} onHide={handleClose}>
        <Modal.Header closeButton>
          <Modal.Title>Modal heading</Modal.Title>
        </Modal.Header>
        <Modal.Body>Woohoo, you're reading this text in a modal!</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose}>
            Close
          </Button>
          <Button variant="primary" onClick={handleClose}>
            Save Changes
          </Button>
        </Modal.Footer>
      </Modal>
      </div>
    </>
  );
}

export default UserDetails;
