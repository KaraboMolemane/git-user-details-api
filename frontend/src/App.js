import "./App.css";
import { Route, Routes } from "react-router-dom";
import Landing from "./components/Landing";
import UserDetails from "./components/UserDetails";

function App() {
  return (
    <div className="App">
      <Routes>
        <Route exact path="/" element={<Landing />} />
        <Route path="/user-details" element={<UserDetails />} />
      </Routes>
    </div>
  );
}

export default App;
