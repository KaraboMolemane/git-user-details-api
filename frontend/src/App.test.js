import { render, screen } from "@testing-library/react";
import App from "./App";
import Landing from "./components/Landing";
import UserDetails from "./components/UserDetails";
import renderer from "react-test-renderer";

test('renders back to Homepage link', () => {
  render(<UserDetails />);
  const linkElement = screen.getByText(/Back to Homepage/i);
  expect(linkElement).toBeInTheDocument();
});

  test('Snapshot test', () => {
    const component = renderer.create(<Landing />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});


  /*
LINKS 
https://youtu.be/KT4dFLrlS7A?si=0gfEGOgAKfQxq_sK
  */