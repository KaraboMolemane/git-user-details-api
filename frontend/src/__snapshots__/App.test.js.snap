// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
<div
  className="container my-5"
>
  <div
    className="p-5 text-center bg-body-tertiary rounded-3"
  >
    <br />
    <h1
      className="text-body-emphasis"
    >
      Search GitHub User App
    </h1>
    <p
      className="col-lg-8 mx-auto fs-5 text-muted"
    >
      This app allows you to search for GitHub users and see the following:
    </p>
    <ul
      className="list-unstyled"
    >
      <li>
        User details - including some of their repos, their profile picture, bio, etc.
      </li>
      <li>
        Repo details - including last commit date, creation date, description, etc.
      </li>
    </ul>
    <div
      className="bd-example"
    >
      <form>
        <div
          className="w-50 p-3"
          style={
            Object {
              "marginLeft": "30%",
            }
          }
        >
          <input
            aria-describedby="emailHelp"
            className="form-control d-inline-flex align-items-center"
            id="userInputWord"
            placeholder="Type in text to search for users"
            type="text"
          />
        </div>
      </form>
      <br />
      <div>
        <div>
          Enter text above to search.
        </div>
      </div>
       
      <br />
      <div
        className="d-inline-flex gap-2 mb-5"
      >
        <button
          className="d-inline-flex align-items-center btn btn-primary btn-lg px-4 rounded-pill"
          data-bs-target="#exampleModalFullscreen"
          data-bs-toggle="modal"
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </div>
    </div>
  </div>
  <div
    aria-hidden="true"
    aria-labelledby="exampleModalFullscreenLabel"
    className="modal fade"
    id="exampleModalFullscreen"
    tabindex="-1"
  >
    <div
      className="modal-dialog modal-fullscreen"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h1
            className="modal-title fs-4"
            id="exampleModalFullscreenLabel"
          >
            Search results for "
            "
          </h1>
          <button
            aria-label="Close"
            className="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
        </div>
        <div
          className="modal-body"
        >
          <div
            className="MuiBox-root css-jbotvb"
          >
            <div
              aria-colcount={5}
              aria-multiselectable={false}
              aria-rowcount={1}
              className="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-5wly58-MuiDataGrid-root"
              role="grid"
            >
              <div
                className="MuiDataGrid-main css-204u17-MuiDataGrid-main"
              >
                <div
                  className="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
                  role="presentation"
                  style={
                    Object {
                      "lineHeight": "56px",
                      "maxHeight": 56,
                      "minHeight": 56,
                    }
                  }
                >
                  <div
                    className="MuiDataGrid-columnHeadersInner css-gl260s-MuiDataGrid-columnHeadersInner"
                    role="rowgroup"
                  />
                </div>
              </div>
              <div
                className="MuiDataGrid-footerContainer MuiDataGrid-withBorderColor css-wop1k0-MuiDataGrid-footerContainer"
              >
                <div />
                <div
                  className="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
                >
                  <div
                    className="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
                  >
                    <div
                      className="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                    />
                    <p
                      className="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                    >
                      0â€“0 of 0
                    </p>
                    <div
                      className="MuiTablePagination-actions"
                    >
                      <button
                        aria-label="Go to previous page"
                        className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                        disabled={true}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={-1}
                        title="Go to previous page"
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="KeyboardArrowLeftIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                          />
                        </svg>
                      </button>
                      <button
                        aria-label="Go to next page"
                        className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                        disabled={true}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onContextMenu={[Function]}
                        onDragLeave={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onKeyUp={[Function]}
                        onMouseDown={[Function]}
                        onMouseLeave={[Function]}
                        onMouseUp={[Function]}
                        onTouchEnd={[Function]}
                        onTouchMove={[Function]}
                        onTouchStart={[Function]}
                        tabIndex={-1}
                        title="Go to next page"
                        type="button"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="KeyboardArrowRightIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <button
            className="btn btn-secondary"
            data-bs-dismiss="modal"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
